listenbrainz/listenstore/timescale_listenstore.py

```diff
@@ -25,59 +25,75 @@
 from listenbrainz.db.dump import SchemaMismatchException
 from listenbrainz.listen import Listen
 from listenbrainz.listenstore import ListenStore
-from listenbrainz.listenstore import ORDER_ASC, ORDER_TEXT, \
-    USER_CACHE_TIME, REDIS_USER_TIMESTAMPS, LISTENS_DUMP_SCHEMA_VERSION
+from listenbrainz.listenstore import ORDER_ASC, ORDER_TEXT, LISTENS_DUMP_SCHEMA_VERSION
 from listenbrainz.utils import create_path, init_cache
 
-REDIS_TIMESCALE_USER_LISTEN_COUNT = "ls.listencount."  # append username
+# Append the user name for both of these keys
+REDIS_USER_LISTEN_COUNT = "lc."
+REDIS_USER_TIMESTAMPS = "ts."
+REDIS_TOTAL_LISTEN_COUNT = "lc-total"
+
 DUMP_CHUNK_SIZE = 100000
 NUMBER_OF_USERS_PER_DIRECTORY = 1000
 DUMP_FILE_SIZE_LIMIT = 1024 * 1024 * 1024  # 1 GB
 DATA_START_YEAR = 2005
-SECONDS_IN_TIME_RANGE = 432000
+DATA_START_YEAR_IN_SECONDS = 1104537600
+
+# How many listens to fetch on the first attempt. If we don't fetch enough, increase it by WINDOW_SIZE_MULTIPLIER
+DEFAULT_FETCH_WINDOW = 30 * 86400  # 30 days
+
+# When expanding the search, how fast should the bounds be moved out
+WINDOW_SIZE_MULTIPLIER = 3
+
+LISTEN_COUNT_BUCKET_WIDTH = 2592000
 
 
 class TimescaleListenStore(ListenStore):
     '''
         The listenstore implementation for the timescale DB.
     '''
 
-    REDIS_TIMESCALE_TOTAL_LISTEN_COUNT = "ls.listencount.total"
-    TOTAL_LISTEN_COUNT_CACHE_TIME = 5 * 60
-    USER_LISTEN_COUNT_CACHE_TIME = 3600  # in seconds. 1 hour
-
     def __init__(self, conf, logger):
         super(TimescaleListenStore, self).__init__(logger)
 
         timescale.init_db_connection(conf['SQLALCHEMY_TIMESCALE_URI'])
 
         # Initialize brainzutils cache
-        self.ns = conf['REDIS_NAMESPACE']
-        init_cache(host=conf['REDIS_HOST'], port=conf['REDIS_PORT'], namespace=conf['REDIS_NAMESPACE'])
-        self.dump_temp_dir_root = conf.get('LISTEN_DUMP_TEMP_DIR_ROOT', tempfile.mkdtemp())
+        init_cache(host=conf['REDIS_HOST'], port=conf['REDIS_PORT'],
+                   namespace=conf['REDIS_NAMESPACE'])
+        self.dump_temp_dir_root = conf.get(
+            'LISTEN_DUMP_TEMP_DIR_ROOT', tempfile.mkdtemp())
+
+    def set_empty_cache_values_for_user(self, user_name):
+        """When a user is created, set the listen_count and timestamp keys so that we
+           can avoid the expensive lookup for a brand new user."""
 
-    def get_listen_count_for_user(self, user_name, need_exact=False):
+        cache.set(REDIS_USER_LISTEN_COUNT + user_name, 0, time=0, encode=False)
+        cache.set(REDIS_USER_TIMESTAMPS + user_name, "0,0", time=0)
```

Comment:
We can probably use hset here.

Reply:
would this be a hash where the key is REDIS_USER_TIMESTAMPS, the field is user_name, and then the value is the 0 or "0,0" etc?

Reply:
I was thinking REDIS_USER_TIMESTAMPS + user_name to be the hash and min and max as fields in it.